src/Simulation/QSharpFoundation/Convert/Convert.qs

```diff
@@ -63,29 +63,42 @@ namespace Microsoft.Quantum.Convert {
     /// Converts a given big integer to an equivalent integer, if possible.
     /// The function returns a pair of the resulting integer and a Boolean flag
     /// which is true, if and only if the conversion was possible.
-    /// # Remarks
-    /// See [C# BigInteger constructor](https://docs.microsoft.com/dotnet/api/system.numerics.biginteger.-ctor?view=netframework-4.7.2#System_Numerics_BigInteger__ctor_System_Int64_) for more details.
     function MaybeBigIntAsInt(a : BigInt) : (Int, Bool) {
-        body intrinsic;
+        let arr = BigIntAsBoolArray(a);
+        if Length(arr) > 64 {
+            return (0, false);
+        }
+
+        mutable val = 0;
+        for i in 0..(Length(arr) - 1) {
+            set val += arr[i] ? 2 ^ i | 0;
+        }
+
+        return (val, true);
     }
 
 
     /// # Summary
     /// Converts a given big integer to an array of Booleans.
     /// The 0 element of the array is the least significant bit of the big integer.
-    /// # Remarks
-    /// See [C# BigInteger constructor](https://docs.microsoft.com/dotnet/api/system.numerics.biginteger.-ctor?view=netframework-4.7.2#System_Numerics_BigInteger__ctor_System_Int64_) for more details.
     function BigIntAsBoolArray(a : BigInt) : Bool[] {
-        body intrinsic;
+        mutable val = a;
```

Comment:
I don't think it's critical, but this will change behaviour. So far the returned array size was always a multiple of 8 due to byte padding of the internal BigInt representation in C#. Now this will not be any longer the case, and in fact behave more closely to how `IntAsBoolArray` behaves.

Reply:
You are totally right, and it broke some of the conversion logic when it came to negative numbers. I've significantly updated the logic to now match the existing behavior and correctly handle padding and two's complement for negative numbers. I included some comments too, to clarify.